src/query/functions/aggregation/take.go

```diff
@@ -0,0 +1,176 @@
+	params NodeParams,
+) (parser.Params, error) {
+	takeTop := opType == TopKType
+	if !takeTop && opType != BottomKType {
+		return baseOp{}, fmt.Errorf("operator not supported: %s", opType)
+	}
+
+	heap := utils.NewFloatHeap(takeTop, int(params.Parameter))
+	fn := func(values []float64, buckets [][]int) []float64 {
+		return takeFn(heap, values, buckets)
+	}
+
+	return newTakeOp(params, opType, fn), nil
+}
+
+// takeOp stores required properties for take ops
+type takeOp struct {
+	params   NodeParams
+	opType   string
+	takeFunc takeFunc
+}
+
+// OpType for the operator
+func (o takeOp) OpType() string {
+	return o.opType
+}
+
+// String representation
+func (o takeOp) String() string {
+	return fmt.Sprintf("type: %s", o.OpType())
+}
+
+// Node creates an execution node
+func (o takeOp) Node(
+	controller *transform.Controller,
+	_ transform.Options,
+) transform.OpNode {
+	return &takeNode{
+		op:         o,
+		controller: controller,
+	}
+}
+
+func newTakeOp(params NodeParams, opType string, takeFunc takeFunc) takeOp {
+	return takeOp{
+		params:   params,
+		opType:   opType,
+		takeFunc: takeFunc,
+	}
+}
+
+// takeNode is different from base node as it only uses grouping to determine
+// groups from which to take values from, and does not necessarily compress the
+// series set as regular aggregation functions do
+type takeNode struct {
+	op         takeOp
+	controller *transform.Controller
+}
+
+// Process the block
+func (n *takeNode) Process(ID parser.NodeID, b block.Block) error {
+	stepIter, err := b.StepIter()
+	if err != nil {
+		return err
+	}
+
+	params := n.op.params
+	meta := stepIter.Meta()
+	seriesMetas := utils.FlattenMetadata(meta, stepIter.SeriesMeta())
+	buckets, _ := utils.GroupSeries(
+		params.MatchingTags,
+		params.Without,
+		n.op.opType,
+		seriesMetas,
+	)
+
+	// retain original metadatas
+	builder, err := n.controller.BlockBuilder(meta, stepIter.SeriesMeta())
+	if err != nil {
+		return err
+	}
+
+	if err := builder.AddCols(stepIter.StepCount()); err != nil {
+		return err
+	}
+
+	for index := 0; stepIter.Next(); index++ {
+		step, err := stepIter.Current()
+		if err != nil {
+			return err
+		}
+
+		values := step.Values()
+		aggregatedValues := n.op.takeFunc(values, buckets)
+		builder.AppendValues(index, aggregatedValues)
+	}
+
+	nextBlock := builder.Build()
+	defer nextBlock.Close()
+	return n.controller.Process(nextBlock)
+}
+
+func takeFn(heap utils.FloatHeap, values []float64, buckets [][]int) []float64 {
+	cap := heap.Cap()
+	for _, bucket := range buckets {
+		// If this bucket's length is less than or equal to the heap's
+		// capacity do not need to clear any values from the input vector,
+		// as they are all included in the output.
+		if len(bucket) <= cap {
+			continue
+		}
+
+		// Add values from this bucket to heap, clearing them from input vector
+		// after they are in the heap.
+		for _, idx := range bucket {
+			val := values[idx]
+			if !math.IsNaN(val) {
+				heap.Push(values[idx], idx)
```

Comment:
I think your logic gets simpler if instead of pushing values to the heap one by one, you heapify with all values at once and extract k values in the end.

Reply:
It's quite difficult to know which values to add to the heap. Each bucket contains the indices that should be used for that takeK function, representing the indices that correspond to the requested grouping by tags. 

I'd have to gather each of the values and its corresponding index that matches the bucket here, then push that vector into the heap; which would still act pretty much the same, so it wouldn't make the logic any simpler

Reply:
You're right, iterating wouldn't change. Iterating over values is simple, I meant you wouldn't need to have a complex Push operation